### DISCUSSION TITLE
YOLOv5: Model head explained

### DISCUSSION THREAD

**Discussion by jeannotdamoiseaux** [ultralytics/yolov5#8281] [Q&A]
Hi, my current understanding of YOLOv5 is that it has three modules: Backbone (CSPDarknet), Neck (PANet) and head (?). The head is a one-stage detector which localizes and classifies the bounding boxes simultaneously. However, I don't seem to find how this final module is constructed. I assume it's again a neural network, but what are the layers? Or how is this module called / how does it work in general?

**Comment by zhiqwang** [comment]
Hi @jeannot-github , 

We restructured the  YOLOv5's model into following four sub-modules in the layout of TorchVision for better understanding. And the refactored network is exactly the same as the official one.

- [BackboneWithPAN](https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d1009d17356723e45d26af94a38485e20d44/yolort/models/backbone_utils.py#L11-L32)
- [YOLOHead](https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d1009d17356723e45d26af94a38485e20d44/yolort/models/box_head.py#L14-L23)
- [PostProcess](https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d1009d17356723e45d26af94a38485e20d44/yolort/models/box_head.py#L360-L369)
- [SetCriterion](https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d1009d17356723e45d26af94a38485e20d44/yolort/models/box_head.py#L85-L101)

The `SetCriterion` is almost same with YOLOv5's [`ComputeLoss`](https://github.com/ultralytics/yolov5/blob/8a66ebad44e8ecf90c7d27757c832579398d4baf/utils/loss.py#L91-L92), the only differences is the output type.

**Reply by SiwarRziguiING3** [reply]
salut @zhiqwang,
vous pouvez expliquer comment ça marche les quatre sous-module? 
et qu'est ce qu'il remplace la fonction maxpooling  ?

**Reply by jeannotdamoiseaux** [reply]
@zhiqwang, I am not sure if you are referring to the YOLOv5 model by Ultralytics or a modified version?

**Reply by zhiqwang** [reply]
> I am not sure if you are referring to the YOLOv5 model by Ultralytics or a modified version?

Hi @jeannot-github , We modified the model but we did use exactly the "same" structure as YOLOv5, and you can also load the checkpoints trained by v5 directly in our project, see https://github.com/zhiqwang/yolov5-rt-stack#loading-checkpoint-from-official-yolov5 for more details.

> vous pouvez expliquer comment ça marche les quatre sous-module?

Hi @SiwarRziguiING3 , This is a long story, In short, this abstraction is widely used in [torchvision](https://github.com/pytorch/vision/tree/main/torchvision/models/detection) and [detectron2](https://github.com/facebookresearch/detectron2) and is also a very widely used solution for building object detection networks.

> et qu'est ce qu'il remplace la fonction maxpooling ?

I don't quite understand this question, which maxpooling are you trying to modify?

**Reply by jeannotdamoiseaux** [reply]
@zhiqwang, so the modules I indicated are wrong or outdated? Because this is how it is described in multiple papers and blogs.

**Reply by zhiqwang** [reply]
> so the modules I indicated are wrong or outdated? Because this is how it is described in multiple papers and blogs.

The head and neck are abstracted by people, not everyone will use a uniform abstraction, such as torchvision and detectron2 they are not applied to an abstraction of Neck.

YOLOv5's backbone is a revised version of DarkNet, and it might look like the following  https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d10/yolort/models/darknetv6.py#L31-L47 if you write it in torchvision layout. It's true YOLOv5's Neck is evolved from PANet, but it's different from the original implementation, and it might look like this https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d10/yolort/models/path_aggregation_network.py#L44-L75  if you write it in torchvision layout.

**Reply by glenn-jocher** [reply]
@jeannot-github YOLOv5 architecture is defined in a model yaml, i.e. for YOLOv5s:
https://github.com/ultralytics/yolov5/blob/a2a1ed201d150343a4f9912d644be2b210206984/models/yolov5s.yaml#L12-L48

**Reply by jeannotdamoiseaux** [reply]
Does anyone have any recommendations (reading, videos, documentation etc.) to understand the C3, SPPF and Detect layers?

**Reply by glenn-jocher** [reply]
@jeannot-github these are all defined in models/common.py and models/yolo.py, i.e.:

https://github.com/ultralytics/yolov5/blob/27d831b6e4ae4b0286ba0159f5c8542e052cd3c9/models/common.py#L209-L225
